File: base/recsys.py
from __future__ import annotations

from typing import TYPE_CHECKING, List, Optional, Any

import abc

if TYPE_CHECKING:
  from base.model import HKModel
  from base.agent import HKAgent

class HKModelRecommendationSystem(abc.ABC):

  def __init__(self, model: HKModel):
    self.model = model

  def post_init(self, dump_data: Optional[Any] = None):
    pass

  def pre_step(self):
    pass

  @abc.abstractmethod
  def recommend(self, agent: HKAgent, neighbors: List[HKAgent], count: int) -> List[HKAgent]:
    pass

  def pre_commit(self):
    pass

  def post_step(self, changed: List[int]):
    pass
  
  def dump(self) -> Any:
    return None

File: ehk-model/model/recsys.go
package model

// HKModelRecommendationSystem defines the interface for recommendation systems
type HKModelRecommendationSystem interface {
	// PostInit is called after the model is initialized
	// dumpData is passed as pointer
	PostInit(dumpData []byte)

	// PreStep is called before each model step
	PreStep()

	// PreCommit is called before changes are committed
	PreCommit()

	// PostStep is called after each model step
	PostStep(changed []*RewiringEventBody)

	// Recommend returns recommendations for an agent
	Recommend(agent *HKAgent, neighbors []*HKAgent, count int) []*TweetRecord

	// Dump returns internal data for debugging/analysis
	Dump() []byte
}

type BaseRecommendationSystem struct {
	// do nothing, provide default empty methods
}

// for type check
func _() HKModelRecommendationSystem {
	return &BaseRecommendationSystem{}
}

func (rs *BaseRecommendationSystem) PostInit(dumpData []byte) {
}

func (rs *BaseRecommendationSystem) PreStep() {
}

func (rs *BaseRecommendationSystem) PreCommit() {
}

func (rs *BaseRecommendationSystem) PostStep(changed []*RewiringEventBody) {
}

func (rs *BaseRecommendationSystem) Recommend(agent *HKAgent, neighbors []*HKAgent, count int) []*TweetRecord {
	return []*TweetRecord{}
}

func (rs *BaseRecommendationSystem) Dump() []byte {
	return nil
}
